src/Interpreters/TLSLog.cpp

```diff
@@ -0,0 +1,234 @@
+#include <Interpreters/TLSLog.h>
+
+#include <chrono>
+#include <memory>
+#include <Columns/ColumnArray.h>
+#include <Common/DateLUT.h>
+#include <Common/Exception.h>
+#include <Common/SipHash.h>
+#include <Core/NamesAndTypes.h>
+#include <Core/Types.h>
+#include <DataTypes/DataTypeArray.h>
+#include <DataTypes/DataTypeDate.h>
+#include <DataTypes/DataTypeDateTime.h>
+#include <DataTypes/DataTypeDateTime64.h>
+#include <DataTypes/DataTypeEnum.h>
+#include <DataTypes/DataTypeLowCardinality.h>
+#include <DataTypes/DataTypeString.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <Interpreters/SystemLog.h>
+#include <Poco/Delegate.h>
+#include <base/scope_guard.h>
+#if USE_SSL
+#include <Poco/Net/X509Certificate.h>
+#include <Poco/Net/SSLManager.h>
+#include <Access/Common/SSLCertificateSubjects.h>
+#include <Poco/Net/VerificationErrorArgs.h>
+#endif
+
+namespace
+{
+using namespace DB;
+
+void fillColumnArray(const Strings & data, IColumn & column)
+{
+    auto & array = typeid_cast<ColumnArray &>(column);
+    size_t size = 0;
+    auto & data_col = array.getData();
+    for (const auto & name : data)
+    {
+        data_col.insertData(name.data(), name.size());
+        ++size;
+    }
+    auto & offsets = array.getOffsets();
+    offsets.push_back(offsets.back() + size);
+}
+
+}
+
+namespace DB
+{
+
+#if USE_SSL
+UInt128 X509CertificateHash(const Poco::Net::X509Certificate &cert)
+{
+    BIO * bio = BIO_new(BIO_s_mem());
+    SCOPE_EXIT({ BIO_free(bio); });
+
+    PEM_write_bio_X509(bio, cert.certificate());
+
+    char * cert_data = nullptr;
```

Comment:
```suggestion
    std::string cert_data;
```

Reply:
I tried this suggestion but hit undefined behaviour : every hash would be different even for strictly equal strings.

Reply:
Yeah looks like it's not as simple as I thought, and we can just pass `&cert_data` to `BIO_get_mem_data`. Probably best to keep the existing impl.